railties/lib/rails/generators/app_name.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module Rails
+  module Generators
+    module AppName # :nodoc:
+      RESERVED_NAMES = %w(application destroy plugin runner test)
+
+      private
+
+        def app_name
+          @app_name ||= original_app_name.tr("-", "_")
+        end
+
+        def original_app_name
+          @original_app_name ||= (defined_app_const_base? ? defined_app_name : File.basename(destination_root)).tr('\\', "").tr(". ", "_")
+        end
+
+        def defined_app_name
+          defined_app_const_base.underscore
+        end
+
+        def defined_app_const_base
+          Rails.respond_to?(:application) && defined?(Rails::Application) &&
+            Rails.application.is_a?(Rails::Application) && Rails.application.class.name.sub(/::Application$/, "")
+        end
+
+        alias :defined_app_const_base? :defined_app_const_base
+
+        def app_const_base
+          @app_const_base ||= defined_app_const_base || app_name.gsub(/\W/, "_").squeeze("_").camelize
+        end
+        alias :camelized :app_const_base
+
+        def app_const
+          @app_const ||= "#{app_const_base}::Application"
+        end
+
+        def valid_const?
+          if /^\d/.match?(app_const)
+            raise Error, "Invalid application name #{original_app_name}. Please give a name which does not start with numbers."
```

Comment:
Curiosity for another time: since 2.6 allows constants to start with non ascii characters, does that mean we could name a Rails app, say, 666?

Reply:
@kaspth not `666`, because it needs to begin with something the unicode database as included in the particular ruby version thinks is an uppercase letter.  Just not limited to ascii.